### PR TITLE
Create a valid cert for the private key used for the packet encryption certificate

### PR DESCRIPTION
**Pull request by aaomidi**
A few concerns I have with this PR immediately:

- I've tested the components separately, but not together in the code. Will need help figuring out how to test them.
- The user key gets stored in `./deploy-tool-output/acme/....`, this is probably not what we want for long term and would like this to just run inside Kubernetes and use k8s' various APIs directly (can configure this pod to have a service account and just use the k8s APIs directly without configuring anything)
-  I'm currently configuring a bunch of stuff for this tool using environment variables. This isn't __terrible__ and might actually be nice once inside a k8s cluster, but it's just something to keep in mind.
- The value that finally ends up in `newCertificates[name]` is going to be something like:

```
-----BEGIN CERTIFICATE-----
cert
-----END CERTIFICATE-----

-----BEGIN CERTIFICATE-----
cert
-----END CERTIFICATE-----

```

From what I read in the documentation this is intended, but I wanted to point it out.

### CODE REVIEW

File: deploy-tool/cert/cert.go
Comment (tgeoghegan) [inline deploy-tool/cert/cert.go:23]: nit: delete blank line

File: deploy-tool/dns/dns.go
Comment (tgeoghegan) [inline deploy-tool/dns/dns.go:28]: nit: just directly `return os.Getenv("D_DNS_PROVIDER")`

File: deploy-tool/cert/cert.go
Comment (tgeoghegan) [inline deploy-tool/cert/cert.go:57]: I think this error string was copy pasted from the one above, but it should refer to the issuance failing.

File: deploy-tool/cert/cert.go
Comment (tgeoghegan) [inline deploy-tool/cert/cert.go:21]: Why use environment variables for this? Even if we decide to go that way, I'd say the main function or module should handle details like that, and this module should take these parameters in the `IssueCertificate` method or perhaps as fields of a struct you define. Doing it this way would make it harder to write tests on this module, because each test function would have to populate the environment appropriately.

File: deploy-tool/main.go
Comment (tgeoghegan) [inline deploy-tool/main.go:186]: "name" here is not what we want to issue the certificate for, but we need to think about how to get the right value. I think we would buy some domain `foo.org`, and then use `${manifestWrapper.KubernetesNamespace}.foo.org`. Of course you'll need to rebase on main after #51 lands to get the version of deploy-tool that plumbs that value in.

File: deploy-tool/dns/dns.go
Comment (tgeoghegan) [inline deploy-tool/dns/dns.go:32]: We gotta talk about how to manage this API token.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: deploy-tool/cert/cert.go
Comment (aaomidi) [reply deploy-tool/cert/cert.go:21]: Yeah, I think that makes sense. I followed the design pattern that Lego had created, but I wasn't really happy with it.

This comes down to my question. How is the user going to execute this? Should it be a CLI based utility with flags? Should it be a configuration file?

File: deploy-tool/dns/dns.go
Comment (aaomidi) [reply deploy-tool/dns/dns.go:32]: Why would we need an API token for the LE ACME endpoint?

File: deploy-tool/main.go
Comment (aaomidi) [reply deploy-tool/main.go:186]: I'll wait on that merge to resolve this issue.

File: deploy-tool/dns/dns.go
Comment (tgeoghegan) [reply deploy-tool/dns/dns.go]: There is a notion of an account and a user in the ACME API: https://tools.ietf.org/html/rfc8555#section-7.3

I think maybe certmagic might be abstracting that away from you, and maybe what they're doing is a brand new user for each certificate request. We should check what's going on, and whether it'd be better to manage a persistent ACME credential (maybe ask the SRE and boulder teams for opinions).

Comment (aaomidi) [reply deploy-tool/dns/dns.go]: Oh you mean the user private key. Yeah since when I wrote this we were planning to only run this on our local machines I assumed we're just going to handle the key.

I think if the intention of this (or another tool that reuses some of the logic) to run automatically in k8s the proper way to do this would be to use the k8s secret storage or just mount it as a volume for the container to see.

File: terraform/modules/data_share_processor/data_share_processor.tf
Comment (aaomidi) [inline terraform/modules/data_share_processor/data_share_processor.tf:1]: Will fix this... was an accident :D

File: deploy-tool/main.go
Comment (aaomidi) [inline deploy-tool/main.go:202]: This might have to be updated to manifestWrapper.KubernetesNamespace + the domain

File: deploy-tool/config/config.go
Comment (tgeoghegan) [inline deploy-tool/config/config.go:24]: `CA` is a little ambiguous. Going by the value in the sample config file, this is the ACME API endpoint, so I think the field should be named accordingly.

File: deploy-tool/config/config.go
Comment (tgeoghegan) [inline deploy-tool/config/config.go:35]: suggest naming this variable `tree` or something

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage.go:133]: Is the creation timestamp going to be the right value if the secret was ever updated by `secretApi.Update`?

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage.go:141]: My guess is that the Kubernetes secret API should provide reasonable ordering guarantees, so I don't think locking is needed. It's worth doing some research on what guarantees the Kubernetes API provides, though.

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage.go:89]: Logging errors to `stderr` is fine. When this runs as a cronjob, GKE will capture container stdout/stderr and forward it into Google's managed logs product.

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage.go:80]: I think you can just `return secretsApi.Delete(...)` and save a few lines here.

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage.go:68]: This should check if `secret.Data` contains `dataKey` and return an error if not.

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [reply deploy-tool/cert/k8_storage.go:28]: Where does the `key` value come from? Do we have any control over it? I'm a little concerned about being clever and trying to sanitize keys because there's a risk of unexpected collisions, though in this case we only ever expect one ACME client credential per Kubernetes cluster so it's a small risk.

File: deploy-tool/cert/cert.go
Comment (tgeoghegan) [reply deploy-tool/cert/cert.go:21]: It seems like all these values now come via config file, so should this comment be updated?

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage.go:102]: What does the `bool` argument we ignore do?

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage.go:37]: Do we need to insert the ACME account ID here, or is that stored separately as a secret by certmagic?

File: deploy-tool/cert/cert.go
Comment (tgeoghegan) [inline deploy-tool/cert/cert.go:48]: Should this specify `SecretStorage` instead of `FileStorage`?

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage.go:15]: `KubernetesSecretStorage` seems like a better name to me. `Secret` is the thing we're storing, but as it is the name doesn't give any insight into _where_ we store the secrets.

File: deploy-tool/cert/k8_storage_test.go
Comment (tgeoghegan) [inline deploy-tool/cert/k8_storage_test.go:35]: I think this test is unnecessary if `IssueCertificate` is updated, right?

File: deploy-tool/dns/dns.go
Comment (tgeoghegan) [inline deploy-tool/dns/dns.go:16]: nit: don't bother defining `cloudflareConfig`, just do `if deployConfig.DNS.CloudflareConfig == nil`

File: deploy-tool/sample.config.toml
Comment (tgeoghegan) [inline deploy-tool/sample.config.toml:5]: Is the indentation significant?

File: deploy-tool/main.go
Comment (tgeoghegan) [inline deploy-tool/main.go:152]: Help me understand where the config file comes from. I don't think any of the values in there are emitted by prior stages of the deployment so I don't think it's emitted from the Terraform module. Looking at what's defined in there, and keeping in mind that imminently we'll be using Google Cloud DNS, it seems like the point of the config file is to allow overriding the default behavior, which would be to use the production Let's Encrypt, use Google Cloud DNS, use Google Cloud Project API credentials from `~/.config/gcloud`, and the Kubernetes driver. I think that's useful for testing this, but now I wonder if in order to absolve every deployer of `prio-server` from being required to make a well-formatted `config.toml`, `deploy-tool/config/config.go` should define default values for the common case.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: deploy-tool/cert/cert.go
Comment (aaomidi) [reply deploy-tool/cert/cert.go:45]: Thanks for catching this - forgot to pipe it through :)

File: deploy-tool/cert/cert.go
Comment (aaomidi) [reply deploy-tool/cert/cert.go:81]: Fair, it's not even really a proper validation. Was just thinking of making it fail faster than it would normally, but the extra code is probably not worth it.

File: deploy-tool/cert/k8_storage.go
Comment (aaomidi) [reply deploy-tool/cert/k8_storage.go:28]: I don't think the long `letsencrypt-acme-client-credential` is a good idea here considering the keys can only be 63 characters long. Do you think we can make that a bit smaller?

The key is essentially a file path from my understanding. It's essentially created by certmagic and in our storage layer we're mapping these file paths to secrets in k8s.

File: deploy-tool/cert/k8_storage.go
Comment (aaomidi) [reply deploy-tool/cert/k8_storage.go:47]: Unfortunately this doesn't describe it well either https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#create-secret-v1-core but just from experience of working with k8s I know a lot of the other commands doesn't like `create` commands if the resource already exists.

File: deploy-tool/main.go
Comment (aaomidi) [reply deploy-tool/main.go:151]: Oh I didn't know about this

File: deploy-tool/cert/k8_storage.go
Comment (aaomidi) [reply deploy-tool/cert/k8_storage.go:133]: Yep an update changes the object entirely I think

File: deploy-tool/cert/k8_storage.go
Comment (aaomidi) [reply deploy-tool/cert/k8_storage.go:102]: It's recursive search. Since technically we don't have a concept of directories, all of our searches are recursive.

Comment (aaomidi) [reply deploy-tool/cert/k8_storage.go:102]: It's not a __big__ deal since we're just returning more information than we would usually.

File: deploy-tool/cert/k8_storage.go
Comment (aaomidi) [reply deploy-tool/cert/k8_storage.go:37]: That's part of the `key`

File: deploy-tool/cert/cert.go
Comment (aaomidi) [reply deploy-tool/cert/cert.go:48]: Actually this is probably something that I should make configurable. Just in case we need to run this locally.

File: deploy-tool/sample.config.toml
Comment (aaomidi) [reply deploy-tool/sample.config.toml:5]: No, it's generally up to us if we want the indentation for sub keys. I've generally seen it without indentation so I'll remove them.

File: deploy-tool/main.go
Comment (aaomidi) [reply deploy-tool/main.go:152]: The config is going to be something the person deploying needs to configure. I'll add a few more defaults but at the very least a few things need to be manually entered:

- Assuming DNS provider defaults to GCP, we'd need the `[dns.gcp]` section (when that gets built out).
- The `acme.email, acme.subscriber_agreement`.
- Defaulting `storage.driver` to `kubernetes` and `storage.kubernetes.namespace` to `prio-server`

File: deploy-tool/cert/k8_storage.go
Comment (tgeoghegan) [reply deploy-tool/cert/k8_storage.go]: Oh, I didn't know there was a length restriction here. I still think `k8s` is redundant, all the moreso if you're trying to make the best use of a limited length. Can we at least put a label or annotation (whichever is more appropriate) with a more descriptive name?

Comment (tgeoghegan) [reply deploy-tool/cert/k8_storage.go]: What I'm reading in k8s docs is that secret names must be "DNS subdomains", which can be "no more than 253" characters, not 63, so it seems like we have enough breathing room to be explicit here.

https://kubernetes.io/docs/concepts/configuration/secret/#overview-of-secrets
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names

File: deploy-tool/config/config.go
Comment (tgeoghegan) [inline deploy-tool/config/config.go:39]: Putting the ACME credential in a Kubernetes namespace seems like a fine idea, but if the idea is that this will eventually be used by a key rotation cronjob running in that same namespace, maybe "key-rotator" would be more appropriate.

File: deploy-tool/cert/k8_storage.go
Comment (aaomidi) [reply deploy-tool/cert/k8_storage.go]: Unfortunately these keys aren't really DNS subdomains, just random strings that we're then going to have to map to DNS subdomains if we need to

So we have the limit defined here https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

I agree though I'll get rid of the `k8s`. These are going to be invisible for us as the users though as it's really only ever going to be modified by the key rotation program

File: deploy-tool/cert/kubernetes_storage.go
Comment (tgeoghegan) [inline deploy-tool/cert/kubernetes_storage.go:87]: Should this be `if !ok`?

File: deploy-tool/cert/cert.go
Comment (tgeoghegan) [inline deploy-tool/cert/cert.go:20]: Why not just set `certmagic.DefaultACME.Agreed = deployConfig.ACME.SubscriberAgreement` below, and then let Boulder or CertMagic throw errors if they don't like the value? Once again, it would be less code and fewer test cases for us to worry about.

File: deploy-tool/main.go
Comment (tgeoghegan) [reply deploy-tool/main.go:152]: I created `enpa-prio-ops@letsencrypt.org` yesterday for another purpose. Can we set that as the default email?

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

**Review by tgeoghegan — APPROVED** [review]

**Pull request merged**
